File: src/lib/db/repositories/inventoryRepository.ts
import { eq, and, inArray, sql } from "drizzle-orm";

import type { AppDatabase } from "../client";
import { inventoryItems, inventorySyncMeta, users } from "../schema";
import type { RollTag, UserInventoryItem } from "../types";
import type { StoredSocketPlug } from "@/lib/inventory/instances/types";

export function upsertInventoryBatch(
  db: AppDatabase,
  userId: number,
  items: UserInventoryItem[],
): void {
  const now = new Date().toISOString();
  db.transaction((tx) => {
    for (const item of items) {
      tx.insert(inventoryItems)
        .values({
          userId,
          instanceId: item.instanceId,
          itemHash: item.itemHash,
          bucket: item.bucket,
          location: item.location,
          characterId: item.characterId ?? null,
          power: item.power,
          isMasterwork: item.isMasterwork ? 1 : 0,
          isCrafted: item.isCrafted ? 1 : 0,
          plugHashes: JSON.stringify(item.plugHashes),
          rollTags: JSON.stringify(item.rollTags),
          statValues: item.statValues ? JSON.stringify(item.statValues) : null,
          gearTier: item.gearTier ?? null,
          socketPlugs: item.socketPlugs ? JSON.stringify(item.socketPlugs) : null,
          syncedAt: item.syncedAt || now,
        })
        .onConflictDoUpdate({
          target: [inventoryItems.userId, inventoryItems.instanceId],
          set: {
            itemHash: sql`excluded.item_hash`,
            bucket: sql`excluded.bucket`,
            location: sql`excluded.location`,
            characterId: sql`excluded.character_id`,
            power: sql`excluded.power`,
            isMasterwork: sql`excluded.is_masterwork`,
            isCrafted: sql`excluded.is_crafted`,
            plugHashes: sql`excluded.plug_hashes`,
            rollTags: sql`excluded.roll_tags`,
            statValues: sql`excluded.stat_values`,
            gearTier: sql`excluded.gear_tier`,
            socketPlugs: sql`excluded.socket_plugs`,
            syncedAt: sql`excluded.synced_at`,
          },
        })
        .run();
    }

    const instanceIds = items.map((i) => i.instanceId);
    if (instanceIds.length === 0) {
      tx.delete(inventoryItems).where(eq(inventoryItems.userId, userId)).run();
    } else {
      tx.delete(inventoryItems)
        .where(
          and(
            eq(inventoryItems.userId, userId),
            sql`${inventoryItems.instanceId} NOT IN (${sql.join(instanceIds.map((id) => sql`${id}`), sql`, `)})`,
          ),
        )
        .run();
    }

    const meta = tx.select().from(inventorySyncMeta).where(eq(inventorySyncMeta.userId, userId)).get();
    const syncVersion = (meta?.syncVersion ?? 0) + 1;
    tx.insert(inventorySyncMeta)
      .values({
        userId,
        itemCount: items.length,
        syncVersion,
        lastFullSyncAt: now,
      })
      .onConflictDoUpdate({
        target: inventorySyncMeta.userId,
        set: {
          itemCount: items.length,
          syncVersion,
          lastFullSyncAt: now,
        },
      })
      .run();

    tx.update(users).set({ lastSyncAt: now }).where(eq(users.id, userId)).run();
  });
}

export function getInventoryStatus(
  db: AppDatabase,
  userId: number,
): { itemCount: number; syncVersion: number; lastFullSyncAt: string | null } | null {
  const meta = db.select().from(inventorySyncMeta).where(eq(inventorySyncMeta.userId, userId)).get();
  if (!meta) return null;
  return {
    itemCount: meta.itemCount,
    syncVersion: meta.syncVersion,
    lastFullSyncAt: meta.lastFullSyncAt,
  };
}

export function queryInventoryByBucket(
  db: AppDatabase,
  userId: number,
  bucket: string,
): UserInventoryItem[] {
  const rows = db
    .select()
    .from(inventoryItems)
    .where(and(eq(inventoryItems.userId, userId), eq(inventoryItems.bucket, bucket)))
    .all();
  return rows.map(rowToItem);
}

export function queryInventoryByHashes(
  db: AppDatabase,
  userId: number,
  itemHashes: number[],
): UserInventoryItem[] {
  if (itemHashes.length === 0) return [];
  const rows = db
    .select()
    .from(inventoryItems)
    .where(and(eq(inventoryItems.userId, userId), inArray(inventoryItems.itemHash, itemHashes)))
    .all();
  return rows.map(rowToItem);
}

export function listInventoryItems(db: AppDatabase, userId: number): UserInventoryItem[] {
  const rows = db.select().from(inventoryItems).where(eq(inventoryItems.userId, userId)).all();
  return rows.map(rowToItem);
}

export function queryInventoryByTags(
  db: AppDatabase,
  userId: number,
  tag: RollTag,
  limit = 20,
): UserInventoryItem[] {
  const rows = db.select().from(inventoryItems).where(eq(inventoryItems.userId, userId)).all();
  return rows
    .map(rowToItem)
    .filter((item) => item.rollTags.includes(tag))
    .slice(0, limit);
}

function rowToItem(row: typeof inventoryItems.$inferSelect): UserInventoryItem {
  const statValues = row.statValues
    ? (JSON.parse(row.statValues) as Partial<Record<string, number>>)
    : undefined;
  return {
    instanceId: row.instanceId,
    itemHash: row.itemHash,
    bucket: row.bucket,
    location: row.location as UserInventoryItem["location"],
    characterId: row.characterId ?? undefined,
    power: row.power,
    isMasterwork: row.isMasterwork === 1,
    isCrafted: row.isCrafted === 1,
    plugHashes: JSON.parse(row.plugHashes) as number[],
    rollTags: JSON.parse(row.rollTags) as RollTag[],
    statValues,
    gearTier: row.gearTier ?? null,
    socketPlugs: row.socketPlugs
      ? (JSON.parse(row.socketPlugs) as StoredSocketPlug[])
      : null,
    syncedAt: row.syncedAt,
  };
}

File: flutter/packages/db/lib/src/repos/inventory_repository.dart
import 'package:drift/drift.dart';

import '../app_database.dart';
import '../tables.dart';
import 'inventory_busy_lock.dart';
import 'inventory_records.dart';
import 'json_codec.dart';

InventoryItemRecord _rowToItem(InventoryItem row) {
  return InventoryItemRecord(
    instanceId: row.instanceId,
    itemHash: row.itemHash,
    bucket: row.bucket,
    location: row.location,
    characterId: row.characterId,
    power: row.power,
    isMasterwork: row.isMasterwork == 1,
    isCrafted: row.isCrafted == 1,
    plugHashes: parseIntJsonArray(row.plugHashes),
    rollTags: parseStringJsonArray(row.rollTags),
    statValues: parseJsonObjectMap(row.statValues),
    gearTier: row.gearTier,
    socketPlugs: parseJsonObjectList(row.socketPlugs),
    syncedAt: row.syncedAt,
  );
}

InventoryItemsCompanion _itemToCompanion(int userId, InventoryItemRecord item) {
  return InventoryItemsCompanion.insert(
    userId: userId,
    instanceId: item.instanceId,
    itemHash: item.itemHash,
    bucket: item.bucket,
    location: item.location,
    characterId: Value(item.characterId),
    power: Value(item.power),
    isMasterwork: Value(item.isMasterwork ? 1 : 0),
    isCrafted: Value(item.isCrafted ? 1 : 0),
    plugHashes: Value(encodeIntJsonArray(item.plugHashes)),
    rollTags: Value(encodeStringJsonArray(item.rollTags)),
    statValues: Value(encodeJsonObjectMap(item.statValues)),
    gearTier: Value(item.gearTier),
    socketPlugs: Value(encodeJsonObjectList(item.socketPlugs)),
    syncedAt: item.syncedAt,
  );
}

/// Full-replace inventory for [userId] in **one** transaction (lock-free).
///
/// Full-replace shape (product `upsertInventoryBatch` intent):
/// 1. Delete existing rows for the user
/// 2. Batch-insert the new set (composite unique `(user_id, instance_id)` enforced)
/// 3. Upsert `inventory_sync_meta` with incremented `syncVersion`
/// 4. Update `users.last_sync_at`
///
/// Returns the new [InventorySyncStatus].
Future<InventorySyncStatus> replaceInventoryBatch(
  AppDatabase db,
  int userId, {
  required List<InventoryItemRecord> items,
  required String now,
}) {
  return db.transaction(() async {
    // 1) Clear prior inventory for this user.
    await (db.delete(db.inventoryItems)..where((t) => t.userId.equals(userId)))
        .go();

    // 2) Batch insert the new set.
    if (items.isNotEmpty) {
      await db.batch((batch) {
        for (final item in items) {
          batch.insert(db.inventoryItems, _itemToCompanion(userId, item));
        }
      });
    }

    // 3) Bump sync meta.
    final prior = await (db.select(db.inventorySyncMeta)
          ..where((t) => t.userId.equals(userId)))
        .getSingleOrNull();
    final nextVersion = (prior?.syncVersion ?? 0) + 1;
    final status = InventorySyncStatus(
      itemCount: items.length,
      syncVersion: nextVersion,
      lastFullSyncAt: now,
    );

    await db.into(db.inventorySyncMeta).insertOnConflictUpdate(
          InventorySyncMetaCompanion(
            userId: Value(userId),
            itemCount: Value(status.itemCount),
            syncVersion: Value(status.syncVersion),
            lastFullSyncAt: Value(status.lastFullSyncAt),
          ),
        );

    // 4) users.last_sync_at
    await (db.update(db.users)..where((t) => t.id.equals(userId))).write(
          UsersCompanion(lastSyncAt: Value(now)),
        );

    return status;
  });
}

/// Full-replace under [lock] (default shared busy lock).
///
/// Throws [InventoryReplaceBusyException] if [userId] already has an exclusive
/// replace in flight.
Future<InventorySyncStatus> replaceInventoryBatchExclusive(
  AppDatabase db,
  int userId, {
  required List<InventoryItemRecord> items,
  required String now,
  InventoryBusyLock? lock,
}) {
  final gate = lock ?? defaultInventoryBusyLock;
  return gate.runExclusive(
    userId,
    () => replaceInventoryBatch(db, userId, items: items, now: now),
  );
}

/// Whether [userId] is busy under the default (or provided) lock.
bool isInventoryReplaceBusy(int userId, {InventoryBusyLock? lock}) {
  return (lock ?? defaultInventoryBusyLock).isBusy(userId);
}

Future<InventorySyncStatus?> getInventoryStatus(
  AppDatabase db,
  int userId,
) async {
  final row = await (db.select(db.inventorySyncMeta)
        ..where((t) => t.userId.equals(userId)))
      .getSingleOrNull();
  if (row == null) return null;
  return InventorySyncStatus(
    itemCount: row.itemCount,
    syncVersion: row.syncVersion,
    lastFullSyncAt: row.lastFullSyncAt,
  );
}

Future<List<InventoryItemRecord>> listInventoryItems(
  AppDatabase db,
  int userId,
) async {
  final rows = await (db.select(db.inventoryItems)
        ..where((t) => t.userId.equals(userId)))
      .get();
  return rows.map(_rowToItem).toList();
}

Future<List<InventoryItemRecord>> queryInventoryByBucket(
  AppDatabase db,
  int userId,
  String bucket,
) async {
  final rows = await (db.select(db.inventoryItems)
        ..where(
          (t) => t.userId.equals(userId) & t.bucket.equals(bucket),
        ))
      .get();
  return rows.map(_rowToItem).toList();
}

Future<List<InventoryItemRecord>> queryInventoryByHashes(
  AppDatabase db,
  int userId,
  List<int> itemHashes,
) async {
  if (itemHashes.isEmpty) return const [];
  final rows = await (db.select(db.inventoryItems)
        ..where(
          (t) => t.userId.equals(userId) & t.itemHash.isIn(itemHashes),
        ))
      .get();
  return rows.map(_rowToItem).toList();
}

Future<List<InventoryItemRecord>> queryInventoryByInstanceIds(
  AppDatabase db,
  int userId,
  List<String> instanceIds,
) async {
  if (instanceIds.isEmpty) return const [];
  final rows = await (db.select(db.inventoryItems)
        ..where(
          (t) => t.userId.equals(userId) & t.instanceId.isIn(instanceIds),
        ))
      .get();
  return rows.map(_rowToItem).toList();
}

/// Filter in memory by roll tag (product scans all user rows).
Future<List<InventoryItemRecord>> queryInventoryByTags(
  AppDatabase db,
  int userId,
  String tag, {
  int limit = 20,
}) async {
  final all = await listInventoryItems(db, userId);
  return all.where((i) => i.rollTags.contains(tag)).take(limit).toList();
}
